SingleSource/UnitTests/Atomic/float_test.cpp

```diff
@@ -0,0 +1,135 @@
+// and the test can compare the two. In reality, some simpler operations (e.g.
+// those conducted through the ALU) can still end up with the correct answer,
+// even when performed nonatomically. These tests do not check the nonatomic
+// result, although it is still outputted to aid in debugging.
+//
+// Each test is performed on all relevant types.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file tests atomic operations on floating point types with aligned
+// memory addresses.
+//
+// The types tested are: float, double.
+// The ops tested are: xchg, cmpxchg.
+//
+//===----------------------------------------------------------------------===//
+
+#include <sys/stat.h>
+
+#include <cstdio>
+#include <thread>
+#include <vector>
+
+#include "numeric.h"
+#include "util.h"
+
+// See numeric.h for an explanation of numeric xchg tests.
+template <typename T>
+void test_float_scalar_xchg() {
+  static constexpr T val = V >> right_shift<T>();
+  static constexpr T expected = val * kExpected;
+  std::vector<std::thread> pool;
+
+  for (int model : atomic_exchange_models) {
+    T afloat = 0;
+    T ffloat = 0;
+    for (int n = 0; n < kThreads; ++n) {
+      pool.emplace_back(looper_numeric_xchg_atomic<T>, &afloat, model);
+    }
+    for (int n = 0; n < kThreads; ++n) {
+      pool[n].join();
+    }
+    pool.clear();
+    for (int n = 0; n < kThreads; ++n) {
+      pool.emplace_back(looper_numeric_xchg_nonatomic<T>, std::ref(ffloat),
+                        model);
+    }
+    for (int n = 0; n < kThreads; ++n) {
+      pool[n].join();
+    }
+    pool.clear();
+    std::cout << "SCALAR (FETCH ADD): "
+              << "atomic: " << afloat << " "
+              << "nonatomic: " << ffloat << "\n";
+    if (lt(afloat, ffloat) || afloat < expected * (1 - kEpsilon) ||
+        afloat > expected * (1 + kEpsilon)) {
+      fail();
+    }
+  }
+}
+
+// See numeric.h for an explanation of numeric cmpxchg tests.
+template <typename T>
+void test_float_scalar_cmpxchg() {
+  static constexpr T val = V >> right_shift<T>();
+  static constexpr T expected = val * kExpected;
+  std::vector<std::thread> pool;
+
+  for (int success_model : atomic_compare_exchange_models) {
+    for (int fail_model : atomic_compare_exchange_models) {
+      T afloat = 0;
+      T ffloat = 0;
+      for (int n = 0; n < kThreads; ++n) {
+        pool.emplace_back(looper_numeric_cmpxchg<T>, &afloat, std::ref(ffloat),
+                          success_model, fail_model);
+      }
+      for (int n = 0; n < kThreads; ++n) {
+        pool[n].join();
+      }
+      pool.clear();
+      std::cout << "SCALAR (FETCH ADD): "
+                << "atomic: " << afloat << " "
+                << "nonatomic: " << ffloat << "\n";
+      if (lt(afloat, ffloat) || afloat < expected * (1 - kEpsilon) ||
+          afloat > expected * (1 + kEpsilon)) {
+        fail();
+      }
+    }
+  }
+}
+
+void test_float() {
```

Comment:
```suggestion
void test_floating_point() {
```
Since one of the types is `float`, `test_float` implies it would only test single precision.